crates/sdk/src/client_cache.rs

```diff
@@ -50,37 +67,166 @@ impl<Row> Default for TableCache<Row> {
     }
 }
 
-impl<Row: Clone + Send + Sync + 'static> TableCache<Row> {
-    /// Apply all the deletes, inserts and updates recorded in `diff`.
-    fn apply_diff(&mut self, diff: &TableUpdate<Row>) {
-        // Apply deletes strictly before inserts,
-        // to avoid duplicates in any unique index.
+type RowEventMap<'r, Row> = HashMap<&'r [u8], &'r Row>;
 
-        for delete in &diff.deletes {
-            self.entries.remove(&delete.bsatn);
-            for index in self.unique_indices.values_mut() {
-                index.remove_row(&delete.row);
-            }
+/// The diff result of applying [`TableUpdate`] to a [`TableCache`].
+pub struct TableAppliedDiff<'r, Row> {
+    deletes: RowEventMap<'r, Row>,
+    inserts: RowEventMap<'r, Row>,
+    update_deletes: Vec<&'r Row>,
+    update_inserts: Vec<&'r Row>,
```

Comment:
Please add a quick comment on these fields (or the struct itself) which hits:
- Whether `deletes` and `update_deletes` are disjoint, and the same for `inserts` and `update_inserts`.
- Whether these are unique- or multi-sets.
- The fact that (if I'm reading correctly) `update_deletes` and `update_inserts` will be empty upon construction, and will be populated by `with_updates_by_pk` or `derive_updates`.